web/src/LogStore.ts

```diff
@@ -0,0 +1,191 @@
+// Client-side log store, which helps client side rendering and filtering of logs.
+//
+// Looslely adapted from the data structures in
+// pkg/model/logstore/logstore.go
+// but with better support for incremental updates and rendering.
+
+type LogSpan = {
+  manifestName: string
+  lastSegmentIndex: number
+  firstSegmentIndex: number
+}
+
+class LogSegment {
+  spanId: string
+  time: string
+  text: string
+  continuesLine: boolean
+
+  constructor(seg: Proto.webviewLogSegment) {
+    this.spanId = seg.spanId || ""
+    this.time = seg.time || ""
+    this.text = seg.text || ""
+    this.continuesLine = false
+  }
+
+  startsLine() {
+    return !this.continuesLine
+  }
+
+  isComplete() {
+    return this.text[this.text.length - 1] == "\n"
+  }
+}
+
+class LogStore {
+  spans: { [key: string]: LogSpan }
+  segments: LogSegment[]
+
+  constructor() {
+    this.spans = {}
+    this.segments = []
+  }
+
+  append(logList: Proto.webviewLogList) {
+    let newSpans = logList.spans as { [key: string]: Proto.webviewLogSpan }
+    for (let key in newSpans) {
+      let exists = this.spans[key]
+      if (!exists) {
+        this.spans[key] = {
+          manifestName: newSpans[key].manifestName || "",
+          firstSegmentIndex: -1,
+          lastSegmentIndex: -1,
+        }
+      }
+    }
+
+    let segments = logList.segments ?? []
+    segments.forEach(segment => {
+      let newSegment = new LogSegment(segment)
+      let span = this.spans[newSegment.spanId]
+      if (!span) {
+        // If we don't have the span for this log, we can't meaninfully print it,
+        // so just drop it. This means that there's a bug on the server, and
+        // the best the client can do is fail gracefully.
+        return
+      }
+      let isStartingNewLine = false
+      if (span.lastSegmentIndex == -1) {
+        isStartingNewLine = true
+      } else if (this.segments[span.lastSegmentIndex].isComplete()) {
+        isStartingNewLine = true
+      }
+
+      if (span.firstSegmentIndex == -1) {
+        span.firstSegmentIndex = this.segments.length
+      }
+      span.lastSegmentIndex = this.segments.length
+
+      newSegment.continuesLine = !isStartingNewLine
+      this.segments.push(newSegment)
+    })
+  }
+
+  allLog() {
+    return this.manifestLog("")
+  }
+
+  manifestLog(mn: string) {
+    let lastLineCompleted = false
+    let allLogs = mn == ""
```

Comment:
should we use `===` here and on the line below?

Reply:
done